crates/cranelift/src/func_environ.rs

```diff
@@ -1681,6 +1682,54 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
         Ok(pos.ins().call_indirect(sig_ref, func_addr, &real_call_args))
     }
 
+    // At this time, this looks a lot like translate_call_indirect.  But, it
+    // will soon change if an unchecked indirect call is added to cranelift, so
+    // when it breaks, just do that instead of factoring it with call_indirect
+    fn translate_call_ref(
+        &mut self,
+        builder: &mut FunctionBuilder,
+        sig_ref: ir::SigRef,
+        callee: ir::Value,
+        call_args: &[ir::Value],
+    ) -> WasmResult<ir::Inst> {
+        let pointer_type = self.pointer_type();
+
+        // Check for whether the callee is null, and trap if so.
+        builder.ins().trapz(callee, ir::TrapCode::NullReference);
```

Comment:
Actually if we wanted to get _super_ clever here (which is totally ok if you'd rather not) I think that this check can be entirely skipped. The `.load()` below will segfault if the `callee` is a null pointer which we can catch in a signal handler to report a trap, which means that technically speaking we can skip the check entirely here. I think the codegen may need to get changed slightly since the `MemFlags::trusted` used when loading the `func_addr` may or may not be right (a Cranelift expert would want to weigh in).

Anyway just a tiny optimization that probably doesn't matter too much in the grand scheme of things, and more than ok to defer to later.

Reply:
I went ahead an opened https://github.com/bytecodealliance/wasmtime/issues/5291 since this applies to `call_indirect` as well.

Reply:
Indeed, we would want to remove the `trusted` flag here, if the load can trap, and this would require a bit more plumbing to emit the right trap code associated with this load's PC (it's no longer a heap-out-of-bounds). But this is probably an O(tens of lines)-sized change, not too bad. Happy to dig a bit and point out specific bits if you want to do this in a followup PR.